### PR TITLE
feat: add custom transformer for fiddle code blocks

### PR DESCRIPTION
**Pull request by DeerMichel**
To provide a seamless way to start fiddles from Electron's docs, it is necessary to store the fiddle's url somewhere. A good place turned out to be next to the language definition of a markdown code block. This maintains GitHub's syntax highlighting as well as hides the url from the user when looking at the file on GitHub.com. Example:
```
```javascript fiddle='electron-fiddle://gist/0cbxyz'
const { app } = require('electron')
...
```

This PR makes use of `hubdown`'s new `runBefore` option to inject a transformer that parses the url from the language definition and puts it into a html data-attribute of the `<code>` tag. The transformer acts in such a way that further custom properties could be added to the language definition at a later time.

cc @felixrieseberg @zeke

### CODE REVIEW

**Comment by DeerMichel** [comment]
umm did CI just die?

image

**Comment by zeke** [comment]
GitHub Actions is probably having downtime...

**Comment by DeerMichel** [comment]
Locally i had to increase the node heap bc it required too much memory - probs node is running out of memory here as well.

**Comment by zeke** [comment]
@DeerMichel do you have a sense of how or where it's running out of memory? The exported module is a big JSON object. We might have to address this by changing the public surface area of this module somehow: multiple individual JSON files, or flat files, etc.

**Comment by DeerMichel** [comment]
Actually i was wrong here, the CI failed indeed bc of the outage. Now we have to address the problem that the highlighting plugin (in a newer version than we used before) no longer accepts `text` and `txt` language definitions which are used in the docs. Do you think i can just write a transformer that throws them away?

**Comment by zeke** [comment]
> `text` and `txt` language definitions which are used in the docs. Do you think i can just write a transformer that throws them away?

That seems reasonable to me. Better yet open a pull request on electron/electron that removes them from the source. 

As I think about this more, I may be remembering that there's some kind of linter in e/e that requires all code blocks to specify a language type? I might be making that up, but otherwise I don't why anyone would need to use `txt` or `text`, as that's what's normally implied by unqualified triple-backticks.

File: script/build-module.js [hunk 1]
```diff
@@ -28,6 +29,25 @@ const categoryNames = {
 }
 const IGNORE_PATTERN = '<!-- i18n-ignore -->'
 
+// remark transformer for 'code' blocks to
+// embed fiddle urls as html attributes
+const fiddleUrls = () => tree => {
```
Comment (HashimotoYT) [inline script/build-module.js]: can be this moved to lib folder?

File: script/build-module.js [hunk 2]
```diff
@@ -16,6 +16,7 @@ const GithubSlugger = require('github-slugger')
 const getIds = require('get-crowdin-file-ids')
 const remark = require('remark')
 const links = require('remark-inline-links')
+const visit = require('unist-util-visit')
```
Comment (HashimotoYT) [inline script/build-module.js]: add to package.json?

**Comment by DeerMichel** [comment]
> 
> That seems reasonable to me. Better yet open a pull request on electron/electron that removes them from the source.
> 
> As I think about this more, I may be remembering that there's some kind of linter in e/e that requires all code blocks to specify a language type? I might be making that up, but otherwise I don't why anyone would need to use `txt` or `text`, as that's what's normally implied by unqualified triple-backticks.

Sounds like the cleanest solution to me (needs to be done for every version we have docs for). cc'ing @codebytere as we already talked about that.

File: script/build-module.js (hunk 1 above)
Comment (DeerMichel) [reply script/build-module.js]: good point 👍

**Comment by zeke** [comment]
> needs to be done for every version we have docs for

What about removing them only from e/e master branch, and also adding a transformer here to remove `txt` from older versions of the docs?

**Comment by DeerMichel** [comment]
Hi, thanks for your questions:

> What about adding verifying if this really Fiddle URL

That's a great idea to guarantee more security. I'll implement that.

> Can this be only parsed Fiddle URL

Sorry i don't exactly get what you mean here. Can you please elaborate a bit about it?

> This should works if this is not in js code block? 

Indeed, fiddles only make sense for javascript blocks; i gonna add a check for `lang == javascript`.

**Comment by HashimotoYT** [comment]
For the second question I'm mean this case:

```
```js fiddle='https://google.com' somethingOther='https://example.com'
const a = 1
const b = 2
console.log('im not fine')
```

Here fiddle url sets first, after goes another url and on exit, I'm getting this:
```html
<pre><code class="hljs language-js" data-fiddle-url="https://google.com&#x27; somethingOther=&#x27;https://example.com">
```

You can see here data-fiddle-url contain fiddle url and another option without separating. In data-fiddle-url potentially should be only fiddle url or nothing, not all line with other potential options.

**Comment by DeerMichel** [comment]
So in my mind this should never happen - thanks for pointing out... let me fix the regex.

**Comment by DeerMichel** [comment]
there you go :)

**Comment by DeerMichel** [comment]
@zeke indeed the `/e/e` docs linter will throw an error when a code block is missing its language definition. Imo that's a good decision to explicitly define the contained lang. 

For this reason, my further approach here would be to add another transformer which removes `text` and `txt` tags. What do you think @HashimotoYT?

**Comment by HashimotoYT** [comment]
@DeerMichel I'm fine with to create a transformer to remove `text` and `txt`. But I agree also here with @zeke to remove these in `e/e` and remove this transformer in some time. We can use some hacky code block what doesn't change anything. Also can you please clarify how you get this error?

**Comment by DeerMichel** [comment]
I'm actually in favor of keeping `text` and `txt` since otherwise we can't check with the docs linter that code blocks have an explicitly denoted language. 

The error will occur when bumping `hubdown` from v2.1 to v2.3. Sth of `hubdown`'s deps has been updated and doesn't ignore [non-supported languages](https://github.com/highlightjs/highlight.js/blob/master/docs/css-classes-reference.rst#language-names-and-aliases) anymore (the highlighter in particular). That's why you see the same error in #775.

**Comment by HashimotoYT** [comment]
@DeerMichel What about this?

Безымянный

**Comment by DeerMichel** [comment]
Awesome, that works - then let me fix it on `/e/e/master` and create a temporary transformer here which transforms `text` and `txt` to `plaintext`.

**Comment by DeerMichel** [comment]
Oki ready to merge ✅

**Comment by zeke** [comment]
Looks good! Is it possible to add some tests to verify that this is working as expected?

**Comment by HashimotoYT** [comment]
Last question, why `package-lock.json` so changed. All deps installed by other PR.

**Comment by DeerMichel** [comment]
@HashimotoYT Good question, i should be able to completely remove it from here. Let me try...

**Comment by DeerMichel** [comment]
@zeke for a test, it would be nice to have at least one fiddle-equipped code example in the docs which is currently not the case. Therefore I'd like to get me unblocked here and add one later in a new PR 😄.

**Pull request merged**